README.md
![Image of Random Color Scheme](https://github.com/bernaferrari/RandomColorScheme/raw/main/assets/readme.png)

# Random Color Scheme

Making a coherent Material Design theme is hard. This is a Flutter plugin that generates a random color pallet based on Material Design while respecting WCAG guidelines.

The idea is for you to (temporarily) replace your `ColorScheme()` with a `randomColorSchemeLight()` or `randomColorSchemeDark()`.
On each run, hot refresh or hot restart, the app is going to look different. The theme is printed in terminal, so you can easily copy and paste back into your project with the colors that you want to stay.

The interactive sample allows you to see how it works and the reasoning behind:

<a href="https://bernaferrari.github.io/RandomColorScheme/"><img src="https://github.com/bernaferrari/RandomColorScheme/raw/main/assets/try_here.png" height="50"/></a>

[![Random Color Scheme](https://github.com/bernaferrari/RandomColorScheme/raw/main/assets/sample_preview.png)](https://bernaferrari.github.io/RandomColorScheme/)

## Usage

In the `pubspec.yaml` of your flutter project, add the following dependency:

[![pub package](https://img.shields.io/pub/v/random_color_scheme.svg)](https://pub.dev/packages/random_color_scheme)

```yaml
dependencies:
  random_color_scheme: ^VERSION
```

In your project, just replace the `ColorScheme.dark(...)` with `randomColorSchemeDark()`.
If you want a light theme, there is a `randomColorSchemeLight()`.

```dart
import 'package:random_colorscheme/random_color_scheme.dart';

Theme(
  data: ThemeData(
    colorScheme: randomColorSchemeDark(),
  ),
  child: MyApp(),
)
```
#### How it works
This started in my [Color Studio project](https://github.com/bernaferrari/color-studio).
I started looking at the Material Design Guidelines on color for both light and dark theme.
After extracting some colors (like the Primary, Secondary and Owl Study), I decided to see how similar were they, in what range are they and how they behave together.
Then, I used [HSLuv](https://www.hsluv.org/) with a random number generator and rules detected from my observation. The final adjustments were made when tweaking the sample.
HSLuv is great because only the Lightness attribute affects WCAG calculated contrast, so all themes are guaranteed to pass
a minimum accessibility threshold.

#### Function listing
- `randomColorSchemeDark(seed: null | int): ColorScheme`
- `randomColorSchemeLight(seed: null | int): ColorScheme`
- `randomColorScheme(seed: null | int, isDark: bool): ColorScheme`

Seed: an integer which guarantees the random function is always going to have the same output (i.e., the same ColorScheme).
It is optional.

## Reporting Issues

If you have any suggestions or feedback, issues and pull requests are welcome.
You can report [here](https://github.com/bernaferrari/RandomColorScheme/issues).

## License

    Copyright 2020 Bernardo Ferrari

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
## [0.1.4] - 03/03/2023.

* Update API.
* primaryVariant -> primaryContainer.
* secondaryVariant -> secondaryContainer.

## [0.1.3] - 03/03/2021.

* Add secondaryVariant.

## [0.1.1] - 23/02/2021.

* Update package description and README images to work in pub.dev.

## [0.1.0] - 23/02/2021.

* Add null-safety.

## [0.0.1] - 04/10/2020.
* Initial release.


test/random_color_scheme_test.dart
import 'dart:ui';

import 'package:flutter_test/flutter_test.dart';
import 'package:random_color_scheme/random_color_scheme.dart';

void main() {
  test('test simple dark scenario', () {
    final colorScheme = randomColorSchemeDark(seed: 0);
    expect(colorScheme.primary, Color(0xff89b9ee));
    expect(colorScheme.primaryContainer, Color(0xff54a0e2));
    expect(colorScheme.secondary, Color(0xfff1a28c));
    expect(colorScheme.secondaryContainer, Color(0xffee7a4b));
    expect(colorScheme.surface, Color(0xff0e0000));
    expect(colorScheme.background, Color(0xff2a4024));

    // unchanged
    expect(colorScheme.onPrimary, Color(0xff000000));
    expect(colorScheme.onBackground, Color(0xffffffff));
    expect(colorScheme.onSurface, Color(0xffffffff));
  });

  test('test simple light scenario', () {
    final colorScheme = randomColorSchemeLight(seed: 0);
    expect(colorScheme.primary, Color(0xff2c589f));
    expect(colorScheme.primaryContainer, Color(0xff1f4178));
    expect(colorScheme.secondary, Color(0xff6a5723));
    expect(colorScheme.secondaryContainer, Color(0xff4f4017));
    expect(colorScheme.surface, Color(0xfffbf4f5));
    expect(colorScheme.background, Color(0xfffeffff));

    // unchanged
    expect(colorScheme.onPrimary, Color(0xffffffff));
    expect(colorScheme.onBackground, Color(0xff000000));
    expect(colorScheme.onSurface, Color(0xff000000));
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:random_color_scheme/random_color_scheme.dart';
import 'package:url_launcher/url_launcher_string.dart';

import 'preview/chat_preview.dart';
import 'preview/social_preview.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Random Color Scheme',
      home: RefreshableHome(),
    );
  }
}

class RefreshableHome extends StatefulWidget {
  const RefreshableHome({Key? key}) : super(key: key);

  @override
  State<RefreshableHome> createState() => _RefreshableHomeState();
}

class _RefreshableHomeState extends State<RefreshableHome> {
  int refreshSeed = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ThemeList(refreshSeed),
      appBar: AppBar(
        backgroundColor: Colors.white,
        title: const Text(
          "Random Color Scheme",
          style: TextStyle(color: Colors.black),
        ),
        actions: [
          IconButton(
            icon: const Icon(
              Icons.code_rounded,
              color: Colors.black,
            ),
            tooltip: "Source",
            onPressed: () async {
              await launchUrlString(
                  "https://github.com/bernaferrari/RandomColorScheme");
            },
          )
        ],
        elevation: 0,
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: () {
          setState(() {
            refreshSeed += 25;
          });
        },
        label: const Text("Refresh"),
        icon: const Icon(Icons.refresh),
        backgroundColor: Colors.red[500],
      ),
    );
  }
}

class ThemeList extends StatelessWidget {
  final int refresh;

  const ThemeList(this.refresh, {Key? key}) : super(key: key);

  Theme customTheme({Widget? child, required int i, required bool isDark}) {
    return Theme(
      data: ThemeData(
        colorScheme: randomColorScheme(seed: i + refresh, isDark: isDark),
        outlinedButtonTheme: OutlinedButtonThemeData(
          style: OutlinedButton.styleFrom(
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(8),
            ),
          ),
        ),
        cardTheme: CardTheme(
          elevation: 0,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
          ),
        ),
      ),
      child: const ThemeItem(),
    );
  }

  @override
  Widget build(BuildContext context) {
    final condition = MediaQuery.of(context).size.width > 950;

    return Padding(
      padding: const EdgeInsets.all(4),
      child: ListView(
        children: [
          Card(
            elevation: 0,
            color: const Color(0xffECFDF5),
            shape: RoundedRectangleBorder(
              side: const BorderSide(width: 2, color: Color(0xffA7F3D0)),
              borderRadius: BorderRadius.circular(8),
            ),
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Text(
                """
This is the sample for a library.
The idea is for you to plug randomColorScheme() into your apps and discover new material themes.""",
                style: Theme.of(context).textTheme.titleSmall,
                textAlign: TextAlign.center,
              ),
            ),
          ),
          for (int i = 0; i < 25; i++)
            Flex(
              direction: condition ? Axis.horizontal : Axis.vertical,
              children: [
                Flexible(
                  flex: condition ? 1 : 0,
                  child: customTheme(
                    child: const ThemeItem(),
                    i: i,
                    isDark: true,
                  ),
                ),
                Flexible(
                  flex: condition ? 1 : 0,
                  child: customTheme(
                    child: const ThemeItem(),
                    i: i,
                    isDark: false,
                  ),
                ),
              ],
            ),
        ],
      ),
    );
  }
}

class ThemeItem extends StatelessWidget {
  const ThemeItem({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Card(
      color: Theme.of(context).colorScheme.background,
      elevation: 0,
      child: Row(
        children: const [
          Expanded(child: SocialPreview()),
          Expanded(child: ChatPreview()),
        ],
      ),
    );
  }
}


example/lib/color_output.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:hsluv/hsluvcolor.dart';

class ColorOutput extends StatefulWidget {
  const ColorOutput({Key? key}) : super(key: key);

  @override
  State<ColorOutput> createState() => _ColorOutputState();
}

class _ColorOutputState extends State<ColorOutput> {
  late int currentSegment = PageStorage.of(context)
          .readState(context, identifier: const ValueKey("Selectable")) ??
      0;

  final Map<int, Widget> children = const <int, Widget>{
    0: Text("HEX"),
    1: Text("RGB"),
    2: Text("HSLuv"),
    3: Text("HSV"),
  };

  void copyToClipboard(BuildContext context, String text) {
    Clipboard.setData(ClipboardData(text: text));

    ScaffoldMessenger.of(context).hideCurrentSnackBar();
    final snackBar = SnackBar(
      content: Text('$text copied'),
      duration: const Duration(milliseconds: 1000),
    );
    ScaffoldMessenger.of(context).showSnackBar(snackBar);
  }

  @override
  Widget build(BuildContext context) {
    final surface = Theme.of(context).colorScheme.surface;
    final background = Theme.of(context).colorScheme.background;

    final lum = HSLuvColor.fromColor(background).lightness;

    final arr = [
      Theme.of(context).colorScheme.primary,
      Theme.of(context).colorScheme.secondary,
      Theme.of(context).colorScheme.surface,
      Theme.of(context).colorScheme.background,
    ];

    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          CupertinoSlidingSegmentedControl<int>(
            children: children,
            backgroundColor:
                Theme.of(context).colorScheme.onBackground.withOpacity(0.20),
            thumbColor: surface,
            onValueChanged: onValueChanged,
            groupValue: currentSegment,
          ),
          const SizedBox(height: 8.0),
          for (int i = 0; i < arr.length; i++)
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 4.0),
              child: OutlinedButton(
                style: OutlinedButton.styleFrom(
                  foregroundColor: arr[i],
                  padding: const EdgeInsets.all(16.0),
                  side: BorderSide(color: arr[i], width: 2.0),
                ),
                child: Text(
                  arr[i].retrieveColorStr(currentSegment),
                  style: TextStyle(
                    color: lum > 50 ? Colors.black : Colors.white,
                  ),
                ),
                onPressed: () {
                  copyToClipboard(
                    context,
                    arr[i].retrieveColorStr(currentSegment),
                  );
                },
              ),
            ),
        ],
      ),
    );
  }

  void onValueChanged(int? newValue) {
    if (newValue == null) {
      return;
    }

    setState(() {
      currentSegment = newValue;
      PageStorage.of(context).writeState(context, currentSegment,
          identifier: const ValueKey("Selectable"));
    });
  }
}

extension on Color {
  String retrieveColorStr(int kind) {
    switch (kind) {
      case 0:
        return "#${value.toRadixString(16).substring(2)}";
      case 1:
        return "R:$red G:$green B:$blue";
      case 2:
        final hsluv = HSLuvColor.fromColor(this);
        return "H:${hsluv.hue.round()} S:${hsluv.saturation.round()} L:${hsluv.lightness.round()}";
      case 3:
        final hsv = HSVColor.fromColor(this);
        return "H:${hsv.hue.round()} S:${(hsv.saturation * 100).round()} V:${(hsv.value * 100).round()}";
      default:
        return "error!";
    }
  }
}


example/lib/preview/chat_preview.dart
import 'package:flutter/material.dart';

class ChatPreview extends StatelessWidget {
  const ChatPreview({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final primary = Theme.of(context).colorScheme.primary;

    return Column(
      children: <Widget>[
        const SizedBox(height: 8),
        Text(
          "Chat",
          style: TextStyle(
              color: primary, fontSize: 22, fontWeight: FontWeight.w800),
        ),
        Text(
          "14 participants",
          style: TextStyle(
              color: primary, fontSize: 14, fontWeight: FontWeight.w700),
        ),
        // ListTile(
        //   title: Text("Bernardo"),
        //   subtitle: Text("What's new?"),
        //   leading: Padding(
        //     padding: const EdgeInsets.all(8),
        //     child: Icon(
        //       Icons.accessibility_new_rounded,
        //       size: 24,
        //       color: primary,
        //     ),
        //   ),
        //   trailing: Icon(
        //     Icons.done,
        //     size: 16,
        //     color: primary,
        //   ),
        // ),
        ListTile(
          title: const Text("Alfred"),
          subtitle: const Text(
            "Here is the weather for today",
            overflow: TextOverflow.ellipsis,
          ),
          leading: Padding(
            padding: const EdgeInsets.all(8),
            child: Icon(
              Icons.wb_sunny_outlined,
              size: 24,
              color: primary,
            ),
          ),
          trailing: Icon(
            Icons.done,
            size: 16,
            color: primary,
          ),
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(
              Icons.graphic_eq,
              color: primary,
            ),
            const SizedBox(width: 8),
            Icon(
              Icons.graphic_eq,
              color: Theme.of(context).colorScheme.secondary,
            ),
          ],
        ),
      ],
    );
  }
}


example/lib/preview/social_preview.dart
import 'package:flutter/material.dart';

import '../color_output.dart';

class SocialPreview extends StatelessWidget {
  const SocialPreview({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: const EdgeInsets.all(8),
      clipBehavior: Clip.antiAlias,
      color: Theme.of(context).colorScheme.surface,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
            child: Text(
              "What's happening",
              style: Theme.of(context)
                  .textTheme
                  .titleLarge
                  ?.copyWith(fontWeight: FontWeight.w900),
            ),
          ),
          Container(
            color: Theme.of(context).colorScheme.onSurface.withOpacity(0.24),
            height: 1,
          ),
          ListTile(
            title: const Text(
              "Trending",
              overflow: TextOverflow.ellipsis,
            ),
            subtitle: const Text(
              "Design updates",
              overflow: TextOverflow.ellipsis,
            ),
            contentPadding: const EdgeInsets.symmetric(horizontal: 8),
            leading: Padding(
              padding: const EdgeInsets.all(8),
              child: Icon(
                Icons.account_balance,
                size: 24,
                color: Theme.of(context).colorScheme.primary,
              ),
            ),
            trailing: Padding(
              padding: const EdgeInsets.only(right: 8),
              child: ElevatedButton.icon(
                style: ElevatedButton.styleFrom(
                  padding: const EdgeInsets.all(16),
                  elevation: 0,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(8),
                  ),
                ),
                icon: const Icon(Icons.share_outlined),
                label: Text(
                  "Export",
                  style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                        color: Theme.of(context).colorScheme.onPrimary,
                      ),
                ),
                onPressed: () {
                  showDialog(
                    context: context,
                    builder: (BuildContext context) {
                      return SimpleDialog(
                        title: const Text('Export (click to copy)'),
                        titlePadding: const EdgeInsets.only(left: 16, top: 16),
                        backgroundColor:
                            Theme.of(context).brightness == Brightness.dark
                                ? const Color(0xff080808)
                                : Colors.white,
                        contentPadding: const EdgeInsets.all(8),
                        children: const <Widget>[ColorOutput()],
                      );
                    },
                  );
                },
              ),
            ),
          ),
          Container(
            color: Theme.of(context).colorScheme.onSurface.withOpacity(0.24),
            height: 1,
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 8),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                _NotButton("Primary", Theme.of(context).colorScheme.primary),
                const SizedBox(width: 8),
                _NotButton(
                  "Secondary",
                  Theme.of(context).colorScheme.secondary,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _NotButton extends StatelessWidget {
  final String text;
  final Color color;

  const _NotButton(this.text, this.color);

  @override
  Widget build(BuildContext context) {
    return Material(
      elevation: 0,
      color: Colors.transparent,
      shape: RoundedRectangleBorder(
        side: BorderSide(color: color),
        borderRadius: BorderRadius.circular(8),
      ),
      child: Padding(
        padding: const EdgeInsets.all(8),
        child: Row(
          children: [
            Icon(
              Icons.border_inner,
              color: color,
              size: 16,
            ),
            const SizedBox(width: 8),
            Text(
              text,
              style: Theme.of(context)
                  .textTheme
                  .bodyMedium
                  ?.copyWith(color: color),
              overflow: TextOverflow.ellipsis,
            ),
          ],
        ),
      ),
    );
  }
}


lib/random_color_scheme.dart
library random_color_scheme;

import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:hsluv/hsluvcolor.dart';

const _kPrimary = 'primary';
const _kPrimaryContainer = 'primaryContainer';
const _kSecondary = 'secondary';
const _kSecondaryContainer = 'secondaryContainer';
const _kSurface = 'surface';
const _kBackground = 'background';

/// Method to generate a random Color Scheme that's either light or dark.
ColorScheme randomColorScheme(
    {int? seed, bool isDark = true, bool shouldPrint = true}) {
  if (isDark) {
    return randomColorSchemeDark(seed: seed, shouldPrint: shouldPrint);
  } else {
    return randomColorSchemeLight(seed: seed, shouldPrint: shouldPrint);
  }
}

/// Generates a random Color Scheme with a dark theme.
/// Properties set: [primary, primaryVariant, secondary, surface, background]
///
/// Input: an optional seed for the Random function.
///
/// Input: an option bool to enable/disable printing to console
/// Output: a random ColorScheme.
ColorScheme randomColorSchemeDark({int? seed, bool shouldPrint = true}) {
  final colors = _getRandomMaterialDark(seed: seed);
  if (shouldPrint) {
    print('''
ColorScheme.dark(
    primary: ${colors[_kPrimary]!.toColor()},
    primaryContainer: ${colors[_kPrimaryContainer]!.toColor()},
    secondary: ${colors[_kSecondary]!.toColor()},
    secondaryContainer: ${colors[_kSecondaryContainer]!.toColor()},
    surface: ${colors[_kSurface]!.toColor()},
    background: ${colors[_kBackground]!.toColor()},
)
''');
  }

  return ColorScheme.dark(
    primary: colors[_kPrimary]!.toColor(),
    primaryContainer: colors[_kPrimaryContainer]!.toColor(),
    secondary: colors[_kSecondary]!.toColor(),
    secondaryContainer: colors[_kSecondaryContainer]!.toColor(),
    surface: colors[_kSurface]!.toColor(),
    background: colors[_kBackground]!.toColor(),
  );
}

/// Generates a random Color Scheme with a light theme.
/// Properties set: [primary, primaryVariant, secondary, surface, background]
///
/// Input: an optional seed for the Random function.
///
/// Input: an option bool to enable/disable printing to console
/// Output: a random ColorScheme.
ColorScheme randomColorSchemeLight({int? seed, bool shouldPrint = true}) {
  final colors = _getRandomMaterialLight(seed: seed);

  if (shouldPrint) {
    print('''
ColorScheme.light(
    primary: ${colors[_kPrimary]!.toColor()},
    primaryContainer: ${colors[_kPrimaryContainer]!.toColor()},
    secondary: ${colors[_kSecondary]!.toColor()},
    secondaryContainer: ${colors[_kSecondaryContainer]!.toColor()},
    surface: ${colors[_kSurface]!.toColor()},
    background: ${colors[_kBackground]!.toColor()},
)
''');
  }

  return ColorScheme.light(
    primary: colors[_kPrimary]!.toColor(),
    primaryContainer: colors[_kPrimaryContainer]!.toColor(),
    secondary: colors[_kSecondary]!.toColor(),
    secondaryContainer: colors[_kSecondaryContainer]!.toColor(),
    surface: colors[_kSurface]!.toColor(),
    background: colors[_kBackground]!.toColor(),
  );
}

/// Generates a random dark theme that tries to be Material complaint.
///
/// PRIMARY COLOR
/// Reason 1:
/// > "fully saturated brand color is applied to the floating action button"
///
/// Reason 2:
/// Looking at Material Colors in HSV:
/// Primary - H: 267 S: 47 V: 99
/// Secondary - H: 174 S: 99 V: 85
/// OWL - H: 345 S: 54 V: 100
/// We get this range: 50 < S < 100 and V > 80 (currently not being used)
///
/// Looking at Material Colors in HSLuv:
/// Primary - H: 281 S: 97 L: 65
/// Secondary - H: 176 S: 100 L: 79
/// OWL - H: 360 S: 100 L: 67
/// We get this range: S > 90 and 65 < L < 85
///
/// Conclusion:
/// Primary has a saturation of [60-100] and lightness of [65-85].
///
/// PRIMARY VARIANT COLOR
/// Same as Primary, but with lightness = lightness - 10.
///
/// SECONDARY
/// Same saturation and lightness as Primary, but hue is different.
/// The difference from Material's default Primary and Secondary is ~105.
/// Therefore, hue = (hue + 90 + random(90)) % 180.
/// It will have a difference between [90, 180].
///
/// BACKGROUND
/// Has a lightness between [0, 25].
///
/// SURFACE
/// Has a lightness between [0, 30]. However, if:
/// abs(surfaceLightness - backgroundLightness) < 5,
/// just make the surfaceLightness = backgroundLightness + 5.
/// Surface's lightness can be lower than background's lightness.
/// That was a deliberate choice.
///
Map<String, HSLuvColor> _getRandomMaterialDark({int? seed}) {
  final rng = math.Random(seed);

  // avoid too similar values between background and surface.
  final backgroundLightness = rng.nextInt(26);
  var surfaceLightness = rng.nextInt(31);
  if ((surfaceLightness - backgroundLightness).abs() < 5) {
    surfaceLightness = backgroundLightness + 5;
  }

  final primaryHue = rng.nextInt(360);
  final primarySaturation = 60 + rng.nextInt(41).toDouble();
  final primaryLightness = 65 + rng.nextInt(21).toDouble();

  final secondaryHue = ((primaryHue + 90 + rng.nextInt(90)) % 360).toDouble();

  return {
    _kPrimary: HSLuvColor.fromHSL(
      primaryHue.toDouble(),
      primarySaturation,
      primaryLightness.toDouble(),
    ),
    _kPrimaryContainer: HSLuvColor.fromHSL(
      primaryHue.toDouble(),
      primarySaturation,
      primaryLightness - 10.0,
    ),
    _kSecondary: HSLuvColor.fromHSL(
      secondaryHue,
      primarySaturation,
      primaryLightness,
    ),
    _kSecondaryContainer: HSLuvColor.fromHSL(
      secondaryHue,
      primarySaturation,
      primaryLightness - 10.0,
    ),
    _kSurface: HSLuvColor.fromHSL(
      rng.nextInt(360).toDouble(),
      rng.nextInt(101).toDouble(),
      surfaceLightness.toDouble(),
    ),
    _kBackground: HSLuvColor.fromHSL(
      rng.nextInt(360).toDouble(),
      rng.nextInt(101).toDouble(),
      backgroundLightness.toDouble(),
    ),
  };
}

/// Generates a random light theme that tries to be Material complaint.
///
/// PRIMARY COLOR
/// Looking at Material Colors in HSV:
/// Primary - H: 265 S: 100 V: 93
/// Secondary - 174 S: 99 L: 85
/// We get this range: S > 90 and V > 80 (currently not being used)
///
/// Looking at Material Colors in HSLuv:
/// Primary - H: 272 S: 100 L: 36
/// Secondary - H: 177 S: 100 L: 79
/// We get this range: S > 90 and 35 < L < 80
///
/// Conclusion:
/// Primary has a saturation of [80-100] and lightness of [25-45].
///
/// PRIMARY VARIANT COLOR
/// Same as Primary, but with lightness = lightness - 10.
///
/// SECONDARY
/// Same saturation and lightness as Primary, but hue is different.
/// The difference from Material's default Primary and Secondary is ~105.
/// Therefore, hue = (hue + 90 + random(90)) % 180.
/// It will have a difference between [90, 180].
///
/// BACKGROUND
/// It is very bright [75-100], saturation stops at 70.
///
/// SURFACE
/// Has a lightness between [25, 45].
///
Map<String, HSLuvColor> _getRandomMaterialLight({int? seed}) {
  final rng = math.Random(seed);

  final primaryHue = rng.nextInt(360);
  final primarySaturation = 80.0 + rng.nextInt(16);
  final primaryLightness = 25 + rng.nextInt(21);

  final secondaryHue = ((primaryHue + 90 + rng.nextInt(90)) % 360).toDouble();

  return {
    _kPrimary: HSLuvColor.fromHSL(
      primaryHue.toDouble(),
      primarySaturation,
      primaryLightness.toDouble(),
    ),
    _kPrimaryContainer: HSLuvColor.fromHSL(
      primaryHue.toDouble(),
      primarySaturation,
      primaryLightness - 10.0,
    ),
    _kSecondary: HSLuvColor.fromHSL(
      secondaryHue,
      primarySaturation,
      primaryLightness.toDouble(),
    ),
    _kSecondaryContainer: HSLuvColor.fromHSL(
      secondaryHue,
      primarySaturation,
      primaryLightness - 10.0,
    ),
    _kSurface: HSLuvColor.fromHSL(
      rng.nextInt(360).toDouble(),
      20.0 + rng.nextInt(81),
      primaryLightness + 45.0 + rng.nextInt(56 - primaryLightness),
    ),
    _kBackground: HSLuvColor.fromHSL(
      rng.nextInt(360).toDouble(),
      rng.nextInt(71).toDouble(),
      primaryLightness + 45.0 + rng.nextInt(56 - primaryLightness),
    ),
  };
}
